### PR TITLE
Feature/fix mipmap issues

### PR DESCRIPTION
**Pull request by Mee-gu**
fix mipmap issues

### CODE REVIEW

File: cocos/renderer/CCTexture2D.cpp [hunk 1]
```diff
@@ -345,45 +345,67 @@ bool Texture2D::initWithMipmaps(MipmapInfo* mipmaps, int mipmapsNum, PixelFormat
...
+        }
 
-    backend::StringUtils::PixelFormat format = static_cast<backend::StringUtils::PixelFormat>(pixelFormat);
-    CCASSERT(format != backend::StringUtils::PixelFormat::NONE, "PixelFormat should not be NONE");
-    
-    textureDescriptor.textureFormat = backend::StringUtils::PixelFormat2TextureFormat(format);
-    CCASSERT(textureDescriptor.textureFormat != backend::TextureFormat::NONE, "TextureFormat should not be NONE");
-    
-    textureDescriptor.compressed = info.compressed;
+        backend::StringUtils::PixelFormat format = static_cast<backend::StringUtils::PixelFormat>(pixelFormat);
```
Comment (minggo) [inline cocos/renderer/CCTexture2D.cpp]: PixelFormat in backend::StringUtils, it is strange.

File: cocos/renderer/backend/StringUtils.cpp [hunk 2]
```diff
@@ -78,10 +78,9 @@ std::string StringUtils::SamplerAddressMode2String(const SamplerAddressMode& add
 
 std::string StringUtils::SamplerDescriptor2String(const SamplerDescriptor& descriptor)
 {
-    std::string samplerInfo = descriptor.mipmapEnabled ? "mipmapEnable":"mipmapDisable";
+    std::string samplerInfo;
     samplerInfo += SamplerFilterType2String(descriptor.magFilter);
     samplerInfo += SamplerFilterType2String(descriptor.minFilter);
-    samplerInfo += SamplerFilterType2String(descriptor.mipmapFilter);
     samplerInfo += SamplerAddressMode2String(descriptor.sAddressMode);
     samplerInfo += SamplerAddressMode2String(descriptor.tAddressMode);
```
Comment (minggo) [inline cocos/renderer/backend/StringUtils.cpp]: Please use `string::append()`, which is more efficient.

File: cocos/renderer/backend/opengl/UtilsGL.cpp [hunk 3]
```diff
@@ -126,50 +126,35 @@ GLint UtilsGL::toGLMagFilter(SamplerFilter magFilter)
...
-                return GL_NEAREST_MIPMAP_NEAREST;
-            case SamplerFilter::DONT_CARE:
-                return GL_LINEAR_MIPMAP_NEAREST;
-            }
-        }
-    }
-    else
+    if (hasMipmaps && !isPow2)
     {
+        cocos2d::log("Change minification filter to either NEAREST or LINEAR since non-power-of-two texture occur in %s %s %d", __FILE__, __FUNCTION__, __LINE__);
```
Comment (minggo) [inline cocos/renderer/backend/opengl/UtilsGL.cpp]: Use cocos2d::LOG instead? cocos2d::log will print the log in release mode too.

**Comment by Mee-gu** [comment]
update review

File: tests/cpp-tests/Resources/Images/nonencryptedAtlas.tps [hunk 4]
```diff
@@ -48,7 +48,7 @@
         <key>textureSubPath</key>
         <string></string>
         <key>textureFormat</key>
-        <enum type="SettingsBase::TextureFormat">pvr2ccz</enum>
+        <enum type="SettingsBase::PixelFormat">pvr2ccz</enum>
```
Comment (minggo) [inline tests/cpp-tests/Resources/Images/nonencryptedAtlas.tps]: It is generated by tool, you can not modify it manually.

Comment (Mee-gu) [reply tests/cpp-tests/Resources/Images/nonencryptedAtlas.tps]: *.tps is a project file generated by TexturePacker.

**Review by minggo — APPROVED** [review]

**Pull request merged**